### ISSUE TITLE
Custom resources don't use icon for file in filesystem

### ISSUE THREAD

**Issue by ghost** [godotengine/godot#32706]
**Godot version:**
3.2 alpha-1 (custom build)
**OS/device including version:**
ArcoLinux
**Issue description:**
Practically all of Godot's resource types use a special icon to help the user determine what sort of file they're dealing with, assuming the file extension isn't enough (and a lot of resources use .tres/.res anyway.)  However, custom resources, though an icon can be supplied for them via `class_name`, do not use that icon in the filesystem, despite it working most everywhere else it would be used.

```
class_name Actor, "res://System/Icons/actor_icon.png" extends Resource
```

![image](https://user-images.githubusercontent.com/28724859/66558878-fc3bd480-eb19-11e9-9612-d8b479a36190.png)
Works in the inspector

![image](https://user-images.githubusercontent.com/28724859/66558912-0b228700-eb1a-11e9-9ae6-47eafa71c00a.png)
And in the Create New Resource dialogue

![image](https://user-images.githubusercontent.com/28724859/66559030-4624ba80-eb1a-11e9-97fc-9c193c3cfcc7.png)
And when creating a new resource in a property

![image](https://user-images.githubusercontent.com/28724859/66559191-900da080-eb1a-11e9-94f1-6f9dd38e04df.png)
But, sadly, just a generic file icon used here, the same icon used for all other custom resources.

Ideally the user would be able to identify the resource, wherever it may be, by the icon, much like they do most other resources; this inconsistency breaks that reliance, and makes it considerably more difficult to determine at a glance what kind of file you're dealing with.

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

If yes, please ensure that an up-to-date Minimal Reproduction Project (MRP) is included in this report (a MRP is a zipped Godot project with the minimal elements necessary to reliably trigger the bug). You can upload ZIP files in an issue comment with a drag and drop.

**Comment by cgbeutler** [comment]
I am seeing this in 3.2.3. I will try to make a minimal reproduction tomorrow.

**Comment by bmolyneaux** [comment]
Here is a reproduction project:
[ResourceIcon.zip](https://github.com/godotengine/godot/files/6176488/ResourceIcon.zip)

my_resource has the icon.png in the Create New Resource dialog, but not in the FileSystem panel.

**Comment by cgbeutler** [comment]
> my_resource has the icon.png in the Create New Resource dialog, but not in the FileSystem panel.

It will only show up in the inspector if the resource is in `tool` mode by putting `tool` at the top of the script.

**Comment by luislodosm** [comment]
Same problem. Godot 3.3

Script:

```
extends Resource
class_name CondicionDatos, "res://motor/iconos/condicion.png"
```

Create new resource window:

<img width="302" alt="Captura de pantalla 2021-05-01 a las 11 10 16" src="https://user-images.githubusercontent.com/63356898/116779355-d3699000-aa6d-11eb-992c-218a1455a84c.png">

Filesystem:

<img width="208" alt="Captura de pantalla 2021-05-01 a las 11 08 53" src="https://user-images.githubusercontent.com/63356898/116779341-b59c2b00-aa6d-11eb-8333-17191ca52189.png">

Filesystem (double window):

<img width="217" alt="Captura de pantalla 2021-05-01 a las 11 12 18" src="https://user-images.githubusercontent.com/63356898/116779396-22afc080-aa6e-11eb-8ccf-b0b6b934a2ca.png">

**Comment by coppolaemilio** [comment]
Would love to see this one implemented! I'm offering a $100 bounty for anyone that can fix/implement this since it is much needed for Dialogic 2.0

**Comment by coppolaemilio** [comment]
There was a PR attempting to fix the issue https://github.com/godotengine/godot/pull/51089 but it was rejected for performance reasons.
And it seems like to fix this we also need to address this one: https://github.com/godotengine/godot/issues/27333

Not sure, just trying to add more information to the issue.

**Comment by MikeSchulze** [comment]
same issue happens on Godot 4.0 beta4

**Comment by coppolaemilio** [comment]
@MikeSchulze As far as I know this was never fixed, so it is still a thing.

**Comment by MikeSchulze** [comment]
> @MikeSchulze As far as I know this was never fixed, so it is still a thing.

Thats bad, it would be nice to have this consistent.
I was looking to get such extra info out of selected TreeItem to calculate my context menus.

**Comment by Atlinx** [comment]
I will try making a new PR with one of the strategies from my original PR, where I attached icon gathering to preview generation. This'll mean icons might not update immediately (requiring an editor restart and maybe even a cache clear), but it'll still be performant and not touch core code.

**Comment by X64X32** [comment]
can confirm on 4.0rc2

**Comment by TomerconDevelopers** [comment]
same issue in 4.0.2 rc1 too

![image](https://user-images.githubusercontent.com/62302722/229282934-4c354864-ee6a-463d-9983-5335d4ebe69b.png)

**Comment by YuriSizov** [comment]
Thanks for your reports, but there is no need to confirm it. It's a missing feature that hasn't been implemented yet.

**Comment by z4gon** [comment]
I'm seeing this issue in this build:
```
Godot Engine v4.3.beta1.official.a4f2ea91a - https://godotengine.org
OpenGL API 4.1 Metal - 88 - Compatibility - Using Device: Apple - Apple M1 Pro
```
![Screenshot 2024-06-07 at 6 34 48 PM](https://github.com/godotengine/godot/assets/4588601/ea3f3ccf-7927-43ed-90bc-63cedd8c8d4c)
![Screenshot 2024-06-07 at 6 34 43 PM](https://github.com/godotengine/godot/assets/4588601/28d08996-6e30-44b5-9191-2345e5fc7a38)
![Screenshot 2024-06-07 at 6 34 04 PM](https://github.com/godotengine/godot/assets/4588601/9874e78f-63ae-4cff-97bd-1011b6925ddb)

**Comment by KoBeWi** [comment]
@z4gon Your resource doesn't have `@icon` defined. What is your expected result in this case?

**Comment by z4gon** [comment]
@KoBeWi oooh, ok so two things:
- I didn't know `@icon` was something that could be defined
- I thought all resources (custom or not) had the cube icon

So, knowing this, I think I would expect custom resources to have the cube icon by default?

**Comment by z4gon** [comment]
@KoBeWi I set an `@icon` to the script, but that didn't update the existing `.tres` files. 
I tried closing the editor and opening again, but still not showing. 
I guess this could be a cache thing? maybe deleting `.godot` fixes that.

Also, this seems to only work if the `@icon` is set on the actual script, it doesn't recognize `@icon` coming from a super class. 
If you have a `ResourceB` that extends `ResourceA`, `@icon` needs to be set in both classes.

**Comment by z4gon** [comment]
Oddly enough, if the resource exports a packed scene, it breaks the icon.

## Exports PackedScene
```gdscript
@icon("res://textures/icons/icon_sword.svg")
class_name MyResource
extends Resource

@export var my_value: int = 0
@export var my_scene: PackedScene
```

![Screenshot 2024-06-08 at 1 08 30 PM](https://github.com/godotengine/godot/assets/4588601/4ad6b0c1-ac27-4ad8-866b-41a3c0df4281)
![Screenshot 2024-06-08 at 1 08 34 PM](https://github.com/godotengine/godot/assets/4588601/a8663963-4916-4a95-8f54-3f8ffb577248)

**Comment by KoBeWi** [comment]
> So, knowing this, I think I would expect custom resources to have the cube icon by default?

I checked that, if a Resource has custom icon, the icon will show in FileSystem, but not in the inspector. Iconless resources use generic file icon, because we don't have custom icon for Resource (the cube is used for Object).

> Noticed that this seems to only work if the @icon is set on the actual script, it doesn't recognize @icon coming from a super class.

That's another bug. You can open a new issue.

> Oddly enough, if the resource exports a packed scene, it breaks the icon.

Also another bug.

**Comment by z4gon** [comment]
> I checked that, if a Resource has custom icon, the icon will show in FileSystem, but not in the inspector. 
I re-checked this, and custom icons for resources show up in the Inspector, but you need to close and re-open the editor for them to show up. I reported the bug here: https://github.com/godotengine/godot/issues/92919

> Iconless resources use generic file icon, because we don't have custom icon for Resource (the cube is used for Object).
Would it be cool to have a generic icon for iconless Resources?

> > Noticed that this seems to only work if the @ICON is set on the actual script, it doesn't recognize @ICON coming from a super class.
> That's another bug. You can open a new issue.
Created bug report here: https://github.com/godotengine/godot/issues/92921

> > Oddly enough, if the resource exports a packed scene, it breaks the icon.
> Also another bug.
I re-checked this in an empty project and couldn't reproduce anymore, I can export a PackedScene and it works fine.
It breaks if the PackedScene also references the same texture of the custom icon, which I think was one of the cases I was testing back then. But I couldn't find a reliable way to reproduce, it's a bit random.

**Comment by laspencer91** [comment]
Seeing the icon in the Inspector, but not in FileSystem. Even after reloading the editor.

![image](https://github.com/godotengine/godot/assets/26290454/8c44e25e-ab2f-4ac5-8bd8-3ec5b8811f4d)

![image](https://github.com/godotengine/godot/assets/26290454/7abc7600-a5f2-4940-aafe-fa44661ae2c2)


**Note:** My icon is a PNG

**Comment by KoBeWi** [comment]
The icons were fixed only in 4.3.

**Comment by z4gon** [comment]
> I re-checked this in an empty project and couldn't reproduce anymore, I can export a PackedScene and it works fine.
> It breaks if the PackedScene also references the same texture of the custom icon, which I think was one of the cases I was testing back then. But I couldn't find a reliable way to reproduce, it's a bit random.

I was able to repro, I made a bug report for it here: https://github.com/godotengine/godot/issues/92942
